docs/RFCS/node_monitor.md

```diff
@@ -0,0 +1,69 @@
+- Feature Name: Node Monitor
+- Status: draft
+- Start Date: 2015-08-20
+- RFC PR:
+- Cockroach Issue: #2149 and #620
+
+# Summary
+
+Add a new `Monitor` service on each node that monitors all the other nodes and
+reports on their current status and health. Based on just a node ID, the
+monitor will report the health of the node. Initially this health will only
+be if the node is dead or not, but will expand to include other factors in the
+future. This will also be the ideal location to add any calculations about
+which node would be best suited to take on a new replica.
+
+This new service will work perfectly with #2153 and #2171
+
+# Motivation
+
+The decisions about when to add/remove replicas for rebalancing and repairing
+require the knowledge about the health of other nodes. There needs to be a
+local source of truth for those decisions.
+
+# Detailed design
+
+## Configuration
+Add a new configuration setting called `timeoutUntilNodeDead` which contains
+the number of seconds after which if a node was not heard from, it is
+considered dead. The default value for this will be 5 minutes.
+
+## Monitor
+Add a new service called `Monitor` that starts when the node is started.
+This new service will run until the stopper is called and have access to
+gossip.
+
+`Monitor` will maintain a map of node IDs to node descriptors and a variety of
+heath statistic about the node. It will also maintain a `lastUpdatedTime` which
+will be set whenever a node descriptor is updated. To maintain this map, a
+callback from gossip for node descriptors will be added. When this
+`lastUpdatedTime` is longer than the `timeoutUnntilNodeDead`, the node is
+considered dead and any replicas on this node may be removed. Note that this
+removal will be done elsewhere.
+
+Monitor will maintain a timespan `timeUntilNextDead` which is calculated by
+taking the nearest `lastUpdatedTime` for all the nodes and adding
+`timeoutUntilNodeDead` and the node ID associated with the timeout.
+
+Monitor will trigger on `timeUntilNextDead` which when triggered checks to see
+if that node has not been updated.
+If the node hasn't been updated, it will mark it as dead.
+Then it will calculate the next `timeUntilNextDead` to wake up the service.
+
+# Drawbacks
+
+Can't think of any right now. Perhaps that we're adding a new service, but it
+should be very lightweight.
+
+# Alternatives
+
+1. Instead of creating a new node monitoring service, add all of this into
+   gossip. Gossip already has most of the node information in it.
+   - Gossip requires a good refactoring and I can see this as one of the first
+   steps to do so. This will mean that you can get node lists from somewhere
+   other than gossip and it will include more details as well. Adding these
+   calculations into gossip seems cumbersome.
```

Comment:
Why is adding to gossip cumbersome? It also seems cumbersome for clients of gossip and the monitor service to have to query two different locations to get the full picture about a node (health + stats). The monitor service can certainly copy this data from gossip, but then we'd have the data duplicated.